### ISSUE TITLE
Address initial jsonata.js test suite errors

### ISSUE THREAD

**Issue by wnm3** [IBM/JSONata4Java#40]
@msmiths has updated the pom.xml to retrieve the jsonata.js tests and has created an AgnosticTestSuite to execute these tests. Many new errors / incompatibilities were created. This issue represents the initial set of changes to address these new problems. Where a particular set of tests are failing because that feature has not yet been implemented, a new issue will be created to track them and that test will be temporarily disabled to allow review of other problems. Separate comments will be created here for the problems that have been fixed under this issue.

**Comment by wnm3** [comment]
Calling a function with just the $ reference like below failed to parse and then failed to execute:
blocks:case005: (  $func := function($arg) {$arg.Account.Order[0].OrderID};  $func($)) returns "order103"

The function call definition did not expect the context reference variable $ so it failed to parse. The MappingExpression.g4 was updated to allow correct parsing making what follows the $ optional:
 | DOLLAR (('.' expr) | (ARR_OPEN expr ARR_CLOSE))?       # context_ref

Correcting this allowed the expression to parse. However, the visitPath was failing with an NPE because the rhsCtx was null so attempting to references this in resolvePath failed.

Adding tests for null rhsCtx and returning the context itself resolved this problem.

**Comment by wnm3** [comment]
Calling an empty expression like "()" doesn't parse.
blocks:case000: "()" expects to return null (** no match**)

The parens declaration failed to allow an empty expression so was changed to make the internal expression optional:
 | '(' (expr (';' expr)*)? ')'                            # parens

This allows the expression to be evaluated and return null (** no match**) as expected.

**Comment by wnm3** [comment]
Calling an empty trailing expression in parens fails to parse
blocks:case002: "(1; 2; 3;)" expects to return 3 but fails to parse

The parens declaration failed to allow a semicolon to end the list of expressions in parens so was changed to allow this:
 | '(' (expr (';' (expr)?)*)? ')'                         # parens
so the expected return value succeeds.

**Comment by wnm3** [comment]
Starting a new expression block (parens) should start a new environment for variable assignment that lasts the lifetime of the block.
blocks:case003: ($a:=1; $b:=2; $c:=($a:=4; $a+$b); $a+$c) returns 10 because when $a is overridden to become 4, making c=6 inside the c block, it does not return to be 1 outside that block, so instead of 1+6=7 we are reporting 4+6=10.

Need to implement a stack of variable and function declarations when entering a block, and popping them when exiting a block. When searching for variables or function declarations, search current and if not found and there is another environment on the stack, pop, search, push to see if the parent block environment has the reference desired.

**Comment by wnm3** [comment]
simple-array-selectors:case005 used a negative index and was returning null (** no match **)
In the visitArray, I had to ensure the group being reviewed in a SelectionArrayNode was an array so the resolvedIndexes can be tested for non-array elements.

**Comment by wnm3** [comment]
** on a null should return null (** no match **)
descendent-operator:case016

Added test for null returnArray when descendants is null for visitDescendant

**Comment by wnm3** [comment]
"foo".**.fud
descendant-operator:case003

foo.**."fud" expected to return [ "hello", "world" ] but was not testing the string "fud" in the path
descendant-operator:case004

Added the ability to transform a StringContext in the visitPath and in the visitDescendants into an IDContext so it would resolve correctly.

**Comment by wnm3** [comment]
Account.Order.Product[$lowercase(Description.Colour) = "purple"][0].Price is returning just 34.45 but the test expects [ 34.45, 34.45 ]
predicates:case003

Not sure why the answer expects two entries, when [0] clearly identifies the first of the resulting 2 cell array. Will open as a new Issue and disable for now.

**Comment by wnm3** [comment]
in in ["hello","world"]
inclusion-operator:case004
"world" in in
inclusion-operator:case005

IMHO, this correctly reports a parse error because in is an operator to find a value in an array per the documentation here: https://docs.jsonata.org/comparison-operators#in-inclusion

If the test were rewritten as "in" in ["hello","world"] we correctly report this as false. Likewise, for "world" in "in" we report false, however, the current parser doesn't even recognize in in as anything worth parsing so it is ignored. We could add specific "rules" for "in in" phrases but this seems like overkill. jsonata.js defines a Strings as follows at https://docs.jsonata.org/expressions
String literals can also be created by enclosing the sequence of characters in quotes. Either double quotes " or single quotes ' can be used, provided the same quote type is used for the start and end of the string literal. So, having just in without quotes is the operator.

**Comment by wnm3** [comment]
dataset7: content.integration.$lowercase(name)
inclusion-operator:case008

We were not expecting a path in the parameter to $lowercase, so we hadn't allowed 2 parameters. This has been fixed.

**Comment by wnm3** [comment]
IGNORED: case003.jsonata
comments:case003 we aren't parsing a function to reference $$ in the argument list
line 24:13 at [@85,698:698=')',<9>,24:13]: mismatched input ')' expecting '.'

**Comment by wnm3** [comment]
FIXED: $trim() on null throws an exception when it should return null (** no match **)
function-trim:case002

Fixed TrimFunction to allow argCount == 0 to return null

**Comment by wnm3** [comment]
FIXED: $fsds
missing-paths:case005

This is an reference to an undefined variable so rather than throw an exception, we return null now

**Comment by wnm3** [comment]
IGNORED: $length(""𝄞💩")
function-length:case016

The emojis are not read correctly in Java so we get a 4 character string rather than a 2 character string and therefore our length is 4 not 2.

**Comment by wnm3** [comment]
DEFERRED: ($twice:=function($f){function($x){$f($f($x))}}; $add3:=function($y){$y+3}; $add6:=$twice($add3); $add6(7))
high-order-functions:case000

This requires lambda functions

**Comment by wnm3** [comment]
FIXED: $sqrt(4) expected 2 but got 2.0
function-sqrt:case000

Need to transform to Long if no fractional part.

**Comment by wnm3** [comment]
FIXED: $lookup(Account.Order.Product.ProductID, "Product Name")
lookup-function:case002
expected null (** no match **) but got an empty array

Fixed to return null if nothing is found.

**Comment by wnm3** [comment]
FIXED: $lookup($$, "gust")
lookup-function:case003
expected "null" but got parse error

Fixed ROOT to allow just "$$" without any path

**Comment by wnm3** [comment]
FIXED: $spread(blah)
function-spread: case002

blah is undefined so the argument becomes null. Added tests to ensure a null response is returned when the argument to the function resolves to null.

**Comment by wnm3** [comment]
FIXED (not not happy with approach): $string($spread(function($x){$x*$x}))
function-spread:case003

The lambda function declaration as a parameter to spread was parsed as null so $string(null) was returning "null". Added check on the type of ExprContext referenced so it would return a "" TextNode. Not sure this makes sense as $spread is supposed to work on objects, and having it applied to a lamba function declaration should probably return null (** no match **)...

            if (argObject == null) {
               ExprContext exprCtx = ctx.exprValues().exprList().expr(0);
               if (exprCtx instanceof Function_declContext) {
                  argObject = new TextNode("");
               }
            }

**Comment by wnm3** [comment]
FIXED: Account.Order.Product.(Price < 30 ? "Cheap")
conditionals:case006

We didn't expect a ternary expression to have only two arguments based on https://docs.jsonata.org/other-operators#--conditional. Regardless, we now support the missing false condition and return null.

**Comment by wnm3** [comment]
FIXED: ["Red"[$$="Bus"], "White"[$$="Police Car"]][0]
conditionals:case001

Needed to add logic to processArrayContent to not add a null to the array as it signifies undefined and we don't store undefined in arrays.

**Comment by wnm3** [comment]
FIXED: $number() with 5 set as context
context:case000

NumberFunction wasn't set up to expect nothing passed as a trigger to read the context. Fixed the test for using the context as input so it is true if there are no arguments passed.

**Comment by wnm3** [comment]
FIXED: ($append($data.a, [2]); $data.a) using binding: {"data": {"a" : [1]}}
function-append:case005

We didn't support data binding in the test rig AgnosticTestSuite so several classes were refactored to allow access to the environment to set the data bindings during the TestCase execution.

**Comment by wnm3** [comment]
FIXED: $trim()
function-trim:case002

We were throwing an exception (which is what try.jsonata.org does) but the test expects null (** no match **) so the TrimFunction was changed to return null.

**Comment by wnm3** [comment]
FIXED: [1..5].("Item " & $string())
context:case002

Changed useContext to be true if there are empty parameters

**Comment by wnm3** [comment]
IGNORED-DEFERRED: $encodeUrlComponent("?x=test")
function-encodeUrl-component:case000-002

We haven't created the function for encodeUrlComponent

**Comment by wnm3** [comment]
FIXED: {"test":()}
object-constructor:case006

We didn't test for the value returned to be null. Now, if null we don't add it to the object.

**Comment by wnm3** [comment]
IGNORED-DEFERRED: Account.Order.Product{$string(ProductID): Price}
object-constructor:case010

We don't expect an object construction as part of the path.

**Comment by wnm3** [comment]
FIXED: $exists($exists)
function-exists:case017

ExistsFunction was only checking for the existence of variables so added a check for declared functions and program functions.

**Comment by wnm3** [comment]
IGNORED: $exists(null)
function-exists:case006

Github repo is down so can't test with try.jsonata.org but local jsonata.js reports null yet the test claims it should report true so suspect test is wrong.

**Comment by wnm3** [comment]
FIXED: $max([])
function-max: case003
$min([])
function-max: case017

We were throwing an exception but should have signaled problems by returning null

**Comment by wnm3** [comment]
FIXED: [-2..2].($*$)
range-operator:case006
FIXED: [false..$x]
range-operator:case020
[1..10000000] ~> $count()
FIXED-KLUDGE? range-operator:case021 -- returned the right number but not in scientific notation which the test case specified so I expand all scientific notation before comparison of Strings in AgnosticTestSuite

I was treating $ and $$ as special variables, but have refactored the context stack to be part of the FrameEnvironment and changed getVariable to set $ and $$ from the context stack.

**Comment by wnm3** [comment]
FIXED: Made changes to use SelectorArrayNode where createSequence() was used in jsonata.js

**Comment by wnm3** [comment]
FIXED function-keys tests 001, 003, 004

**Comment by wnm3** [comment]
FIXED comparison-operators tests 010, 027

**Comment by wnm3** [comment]
FIXED function-sum:case002

Added number formatter for 2 decimal places as default for String concatenation.

**Comment by wnm3** [comment]
FIXED function-reduce:case006
FIXED function-reduce:case007

Was throwing an exception rather than returning null for value that evaluates to null
Needed to test for declared functions before standard functions so you could override standard functions.

**Comment by wnm3** [comment]
There was a bug in how I was processing the Content_refContext. I needed to replace the child0 in the context being visited, and not in the original context because the stack can change content on successive calls.

**Comment by wnm3** [comment]
DEFERRED: added SKIP_TEST for all Lambda functions

**Comment by wnm3** [comment]
FIXED: (  $seq := 1;  $reduce($seq, function($x){$x}))
hof-reduce:case008
Added a check for the function signature to ensure at least 2 variables are specified in the variable list. Now throws the exception: "The second argument of reduce function must be a function with at least two arguments" expected if less than 2 arguments

**Comment by wnm3** [comment]
FIXED: (  $concat := function($s){function($a, $b){$string($a) & $s & $string($b)}};  $comma_join := $concat(' ... ');  $reduce([1,2,3,4,5], $comma_join))      
hof-reduce:case001

I had to expand the FrameEnvironment to track named declared functions as well as variables referencing either named declared functions or Jsonata functions as exhibited by the hof-reduce:case001

**Comment by wnm3** [comment]
FIXED: added URL encode and decode functions

Note: AgnosticTestSuite only tested for EvaluateException or ParseException, but I throw an EvaluateRuntimeException (to avoid having to explicitly catch exceptions for functions) so I updated the AgnosticTestSuite to allow this:
```
} catch (Throwable ex) {
   if (testCase.expectParseOrEvaluateException && (ex instanceof ParseException || ex instanceof EvaluateException || ex instanceof EvaluateRuntimeException)) {
      notifier.fireTestFinished(testCase.getDescription());
   } else {
      notifier.fireTestFailure(new Failure(testCase.getDescription(), ex));
   }
}
```

**Comment by wnm3** [comment]
FIXED: added TypeFunction for function-typeOf tests

**Comment by wnm3** [comment]
FIXED: function-power:case001 $power(4, 0.5)
FIXED: function-power:case005 $power(-2, 1/3)

Checked for whole number returns and changed to LongNode
Improved testing for NaN and Infinity to return proper error message

**Comment by wnm3** [comment]
FIXED: $sum($split("12345", "").$number($))
function-split:case004

Needed to allow for argument count test to take into account that we are reading the arguments from the call stack using the context reference.

**Comment by wnm3** [comment]
DEFERRED: tail-recursion tests
Need to implement time and count in AgnosticTestSuite

**Comment by wnm3** [comment]
DEFERRED: transforms
Not implemented yet. See #47

**Comment by wnm3** [comment]
FIXED: $substringAfter("Hello World", "ld")
function-substringafter

Needed to take length of search string into account and return empty string if we are at the end of the string.

**Comment by wnm3** [comment]
FIXED: $boolean(-0.5) to return true
function-boolean

Needed to allow double value tests

**Comment by wnm3** [comment]
DEFERRED: Context variable binding (@) 
Not implemented see #48

See https://docs.jsonata.org/path-operators#-context-variable-binding

**Comment by wnm3** [comment]
DEFERRED: Number Formatting
Not implemented see #49

See https://docs.jsonata.org/numeric-functions#formatnumber

**Comment by wnm3** [comment]
FIXED: function-pad:case0010

Had to address unicode when considering lengths.

**Comment by wnm3** [comment]
FIXED: $join([1,2,3], ", ")
function-join:case009

Needed to check for isTextual rather than isValueNode for elements of the array

**Comment by wnm3** [comment]
( $foo := "defined"; ( $foo := nothing; $foo ) )
variables:case011

Because variables can now hold null values, we need to check whether the current _variableMap contains the variable key to allow null to be returned.

**Comment by wnm3** [comment]
FIXED: [1, $two, ["three", $four]]
array-contructor:case006
FIXED: ["foo.bar", foo.bar, ["foo.baz", foo.blah.baz]]
array-constructor:case007
FIXED: foo.blah.baz.[fud, fud]
array-constructor:case011

Added tests for first and last step references to SelectorArrayNode to use groups, and added logic to track array construction to append vs add arrays

**Comment by wnm3** [comment]
DEFERRED: # Positional  Variable Binding
See Issue #50

**Comment by wnm3** [comment]
DEFERRED: expr": "$.{ type: $average(value), kind: $sum(value) }
object-constructor:case023

See Issue #51

Need to add fieldExpr to MappingExpressions.g4 to recognize the entire expression. At present tree only has the $ and the rest is ignored.

**Comment by wnm3** [comment]
FIXED: function-string:case023

Needed to process additional parameter for prettify option

**Comment by wnm3** [comment]
FIXED: $number()
context:case000 using data_0.json

Needed to remove the test for ctx.parent() == null when determining to use context.

**Comment by wnm3** [comment]
$string(["string", 5], true)
FIXED: function-string:case027

Needed custom pretty printer for array indentation and newlines

**Comment by wnm3** [comment]
FIXED double tests to take into account Infinity and NaN values

**Comment by wnm3** [comment]
$string(1e21)
FIXED: function-string:case027

Needed to emulate JavaScript scientific notation for Java

**Comment by wnm3** [comment]
FIXED: $string({  "string": "hello",  "number": 78.8 / 2,  "null":null,  "boolean": false,  "function": $sum,  "lambda": function(){true},  "object": {    "str": "another",    "lambda2": function($n){$n}  },  "array": []}, false)
function-string:case027

Updated Object construction to address function references as empty strings.

**Comment by wnm3** [comment]
FIXED: determining when to use context variable as the first argument for functions.

**Comment by wnm3** [comment]
DEFERRED: function-tomillis

Need to review the formatting options allowed. See #52

**Comment by wnm3** [comment]
DEFERRED: partial-applications

Need to implement this  feature using ? as a function variable placeholder. See #53

**Comment by wnm3** [comment]
FIXED: $average(1)
function-average:case007

Added logic to return long for whole number values

**Comment by wnm3** [comment]
hof-zip-map: case000 and case001 fail because the array build with zip has an extra array wrapper, so it isn't presenting arrays of numbers to the $sum function.

**Comment by wnm3** [comment]
DEFERRED: 55=>5 won't parse but doesn't throw a parse error either See #59

**Comment by wnm3** [comment]
FIXED: $formatBase(99.5, 2.5)
function-formatBase:case006

Had to add rounding (even down, odd up) to both the number and radix presented.

**Comment by wnm3** [comment]
( $square := function($x){$x*$x}; [1..5] ~> $map($square) ~> $sum() )
function-applications:case011

Needed to check for Fct_chainContext as the parent to use context.

**Comment by wnm3** [comment]
FIXED: BasicExpressionTest simpleTest("[{\"b\":1}][a or true]", "{\"b\":1}"); // jsonata 1.8.2 no longer returns null);

**Comment by wnm3** [comment]
DEFERRED: $0 should throw a parse exception since variable names can not start with a number

See issue #64 
Commented out test in InvalidSyntaxTest

**Comment by wnm3** [comment]
Commented out segment in pom.xml to enable jar files to be built

**Comment by wnm3** [comment]
1.4.2 Release incorporates these fixes
